File: app/src/main/java/io/trewartha/positional/position/CoordinatesDegreesMinutesSecondsFragment.java
package io.trewartha.positional.position;

import android.location.Location;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import io.trewartha.positional.R;

public class CoordinatesDegreesMinutesSecondsFragment extends CoordinatesFragment {

    private TextView latitudeTextView;
    private TextView longitudeTextView;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.coordinates_degrees_minutes_seconds_fragment, container, false);
        latitudeTextView = view.findViewById(R.id.degrees_minutes_seconds_latitude_text_view);
        longitudeTextView = view.findViewById(R.id.degrees_minutes_seconds_longitude_text_view);
        setCoordinates(latitude, longitude);
        return view;
    }

    @Override
    public void setCoordinates(double latitude, double longitude) {
        if (latitudeTextView != null && longitudeTextView != null) {
            latitudeTextView.setText(getLatitudeAsDMS(latitude));
            longitudeTextView.setText(getLongitudeAsDMS(longitude));
        }
    }

    @NonNull
    private String getLatitudeAsDMS(double latitude) {
        String latitudeDMS = Location.convert(latitude, Location.FORMAT_SECONDS);
        latitudeDMS = replaceDelimiters(latitudeDMS);
        if (latitude >= 0.0) {
            latitudeDMS = latitudeDMS + " N";
        } else {
            latitudeDMS = latitudeDMS.replaceFirst("-", "") + " S";
        }
        return latitudeDMS;
    }

    @NonNull
    private String getLongitudeAsDMS(double longitude) {
        String longitudeDMS = Location.convert(longitude, Location.FORMAT_SECONDS);
        longitudeDMS = replaceDelimiters(longitudeDMS);
        if (longitude >= 0.0) {
            longitudeDMS = longitudeDMS + " E";
        } else {
            longitudeDMS = longitudeDMS.replaceFirst("-", "") + " W";
        }
        return longitudeDMS;
    }

    @NonNull
    private String replaceDelimiters(@NonNull String string) {
        return string.replaceFirst(":", "° ").replaceFirst(":", "' ") + "\"";
    }
}

File: app/src/main/java/io/trewartha/positional/position/CoordinatesDegreesMinutesSecondsFragment.kt
package io.trewartha.positional.position

import android.location.Location
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import io.trewartha.positional.R
import kotlinx.android.synthetic.main.coordinates_degrees_minutes_seconds_fragment.*

class CoordinatesDegreesMinutesSecondsFragment : CoordinatesFragment() {

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.coordinates_degrees_minutes_seconds_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setCoordinates(latitude, longitude)
    }

    override fun setCoordinates(latitude: Double, longitude: Double) {
        latitudeTextView?.text = getLatitudeAsDMS(latitude)
        longitudeTextView?.text = getLongitudeAsDMS(longitude)
    }

    private fun getLatitudeAsDMS(latitude: Double): String {
        var latitudeDMS = Location.convert(latitude, Location.FORMAT_SECONDS)
        latitudeDMS = replaceDelimiters(latitudeDMS)
        if (latitude >= 0.0) {
            latitudeDMS += " N"
        } else {
            latitudeDMS = latitudeDMS.replaceFirst("-".toRegex(), "") + " S"
        }
        return latitudeDMS
    }

    private fun getLongitudeAsDMS(longitude: Double): String {
        var longitudeDMS = Location.convert(longitude, Location.FORMAT_SECONDS)
        longitudeDMS = replaceDelimiters(longitudeDMS)
        if (longitude >= 0.0) {
            longitudeDMS += " E"
        } else {
            longitudeDMS = longitudeDMS.replaceFirst("-".toRegex(), "") + " W"
        }
        return longitudeDMS
    }

    private fun replaceDelimiters(string: String): String {
        return string.replaceFirst(":".toRegex(), "° ").replaceFirst(":".toRegex(), "' ") + "\""
    }
}
